### PR TITLE
ARROW-8002: [C++][Dataset][R] Support partitioned dataset writing

### PR DESCRIPTION
**Pull request by bkietz**

### CODE REVIEW

File: cpp/src/arrow/dataset/file_base.cc [hunk 1]
```diff
@@ -154,52 +157,101 @@ FragmentIterator FileSystemDataset::GetFragmentsImpl(
...
+
+  /// The FileSystem and base directory into which fragments will be written
+  std::shared_ptr<fs::FileSystem> filesystem;
+  std::string base_dir;
+
+  /// Batches to be written
+  std::shared_ptr<RecordBatchReader> batches;
+
+  /// An Expression already satisfied by every batch to be written
+  std::shared_ptr<Expression> partition_expression;
```
Comment (jorisvandenbossche) [inline cpp/src/arrow/dataset/file_base.cc:179]: What is this used for?

File: cpp/src/arrow/dataset/filter.cc [hunk 2]
```diff
@@ -1493,5 +1495,191 @@ Result<std::shared_ptr<Expression>> Expression::Deserialize(const Buffer& serial
...
+  RETURN_NOT_OK(offset_builder.Finish(&offsets));
+  return offsets;
+}
+
+// Helper for simultaneous dictionary encoding of multiple arrays.
+//
+// The fused dictionary is the Cartesian product of the individual dictionaries.
+// For example given two arrays A, B where A has unique values ["ex", "why"]
+// and B has unique values [0, 1] the fused dictionary is the set of tuples
+// [["ex", 0], ["ex", 1], ["why", 0], ["ex", 1]].
```
Comment (jorisvandenbossche) [inline cpp/src/arrow/dataset/filter.cc:1522]: What's the behaviour for nulls?

(edit: ah, I see that is checked and raises a NotImplemented error for now)

**Review by jorisvandenbossche — COMMENTED** [review]
Quickly browsed through. 
I find it a bit strange to have writing tied to the Dataset object, but I assume the intent is that as a user you would create a InMemoryDataset from your arrow table to be able to write it?

File: cpp/src/arrow/dataset/file_base.cc (hunk 1 above)
Comment (bkietz) [reply cpp/src/arrow/dataset/file_base.cc:179]: This stores partition information which was already attached at the fragment level, and therefore may reference virtual columns absent from `batches`. Without this we'd need to materialize those columns then waste time Grouping their (identical) entries, finally dropping them in the common case where we don't write them to disk.

File: cpp/src/arrow/dataset/file_base.cc [hunk 3]
```diff
@@ -154,52 +157,101 @@ FragmentIterator FileSystemDataset::GetFragmentsImpl(
...
-      ARROW_ASSIGN_OR_RAISE(auto write_task,
-                            plan.format->WriteFragment({path, filesystem}, input_fragment,
-                                                       scan_options, scan_context));
-      task_group->Append([write_task] { return write_task->Execute(); });
-
-      ARROW_ASSIGN_OR_RAISE(
-          auto fragment, plan.format->MakeFragment(
-                             {path, filesystem}, input_fragment->partition_expression()));
-      fragments.push_back(std::move(fragment));
+  std::string basename;
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/dataset/file_base.cc:163]: Will do.

**Review by nealrichardson — APPROVED** [review]
There will be followups but this is a good start

**Pull request closed without merging**